### ISSUE TITLE
Make output of detekt serializable

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#6879]
I'm not even sure if `Finding2` should have an `Entity`. I think that it would be great to make `Finding2` serializable some how (json, sarif...). And `Entity` has a refenrece to a `KtElement` and that can not be serializable.

I think that this change would be good because:

- The reporters shouldn't work with the complexity of `Entity`. We should need to expand `Finding` to contain the locations and those things but that's it.
- We could change our gradle plugin and have one task to run detekt and another to generete the reports. This would make really fast the use case of execute detekt and then execute detektBaseline.
- Merge reports would be way, way easier. You would get a `List<Detektion>` and then you decide how to merge them but you have all the information there.

The cons:
- Right now the html report uses `Entity` to generate the code snippets. If we don't provide the `Entity` we should look for other ways to implement that.

**Comment by marschwar** [comment]
> * Right now the html report uses `Entity` to generate the code snippets.

Same for markdown

**Comment by BraisGabin** [comment]
I'm working on this now.

The main blocker for making `Detektion` serializable is that `Issue` contains an `Entity` and `Entity` contains a reference to a `KtElement` and the `KtElement` is not serializable.

The first step is to create a new `Entity` class for `Issue`. Right now `CodeSmell` and `Issue` share the same implementation of `Entity` but that should change because we need to remove `KtElement` from `Issue` but we can't do that for `CodeSmell`. At the `CodeSmell` level we need the `KtElement` to know if we should suppress for a `@Suppress` annotation and other similar use cases.

For that reason I started simplifying `Entity`: #7470

Also we need to change the `Location` class inside `Entity`.

At the level of `CodeSmell` we always want to play with absolute paths. The absolute paths are less error prone when you want to access the files. But at the `Issue` level we want to work only with relative paths. The `Issue` is the element that will be serialized so it should be relative to work well with the gradle cache. Also the reports that we generate (html, xml...) should be also relative paths so this change should make those less error prone. But the console output should still be absolute. There are a lot of things to tune.

You should expect a lot of PRs improving our tests to ensure that the paths that we show are the correct ones.

**Comment by BraisGabin** [comment]
Ok, I have the first part of this ready. I was able to remove KtElement and make `location.path` relative keeping all working as before. I can't launch yet the PRs because there are already a lot of PRs open with lots of dependencies.

Once I get this first part merged the second part is to make `Issue` serializable but that should be easy. I vote to use `json` for that serialization but I accept other ideas.

**Comment by cortinico** [comment]
> I vote to use `json` for that serialization but I accept other ideas.

Json sounds fine to me 👍 human readable and we can use kotlinx serialization for it

**Comment by BraisGabin** [comment]
Ok. Let's focus here. To finish this issue we need to be able to Serialize `Detektion`. This is what `Detektion` has:

```kotlin
interface Detektion {
    val issues: List<Issue>
    val rules: List<RuleInstance>
    val notifications: Collection<Notification>
    val metrics: Collection<ProjectMetric>
    val userData: MutableMap<String, Any?>
}
```

I see different problems that need to be solved before making this possible:

1. `userData`: #8556 makes us the life a bit easier because `UserDataHolder` was nearly impossible to serielize but we aren't there yet. `Any` is the problem. I see different solutions:
   - We could use `kotlinx.serialization` to Serialize `Detektion` (to me is the best option). If we do so we could force that the users can add any type to `userData` but it should be marked as `@Serializable`. If it's not it will fail at runtime. I don't like the runtime errors but I don't know how can we give the flexibility of `Any` otherwise. (I didn't prove this, I think it's doable)
   - Change `userData` to be `MutableMap<String, String>`. Serialize a `String` is trivial and if you want to add a complex value you can serialize it to `json` or whatever before you add it to `userData`
2. `notifications`. The notifications are really easy to parse BUT right now the reporters are using this API to do this: `result.add(SimpleNotification("Successfully generated ${report.id} at ${filePath.toUri()}"))`. That's a problem because we want to serialize the output of detekt BEFORE we execute the reporters. So even if that could be done it would be pointless because the other reporters wouldn't see those notifications. We need to refactor that code. Options:
   - Don't print the urls of each report. Right now we don't do it by default, you need to enable `NotificationReport`.
   - Add a configuration on our cli and gradle plugin so the user can decide

I'm going to create a new issue to fix point 2 for now because it's the biggest problem right now.

**Comment by BraisGabin** [comment]
I would like to know your opinion @detekt/maintainers. I want to use kotlinx.serializer to generate a json for `Detektion` and to get a `Detektion` from a json. To do that I have different options:

1. Annotate `Detektion` with `@Serialization`
2. Create DTOs for the serialization.

The first is easier, way easier. But I'm a bit afraid with the future compatibility. The second option needs more code but we would have full control over the serialization and we will be free to change `Detektion` without breaking the serialization.

Then there are greys here. kotlinx.serialization allows you to add serializers `@Serializable(with = ...)` so we could still annotate `Detektion` with the same flexibility. But we will need to maintain the serializer, it will not be free.

What do you think?

**Comment by schalkms** [comment]
Before I comment with my final conclusion, I would like to know the reason for this feeling regarding option 1).
> But I'm a bit afraid with the future compatibility.

**Comment by TWiStErRob** [comment]
I don't think it's a good idea to expose the serialization as I would assume it's an implementation detail of core / gradle plugin? The code generated by the kotlinx compiler plugin becomes part of our public API.

Note: I would write a JSON-schema and serialize into that format, but not expose that serialization to the users, or do expose, but then make it a separate independent binary?

**Comment by BraisGabin** [comment]
The JSON-schema have sense.

The serialization shouldn't be public for the end user but it should be public for the integrators. So, probably you are right. It should not be inside `detekt-api` at all. But it should be inside `detekt-tooling` that it is our api for the integrators.

**Comment by BraisGabin** [comment]
> I would like to know the reason for this feeling regarding option 1).

The problem with the compatibility is that if we decide to add/remove a field in `Detektion` then the serialization changes so the previously generated JSONs may fail to serialize to the new `Detektion` version. Probably this isn't a huge problem because `Detektion` is part of our public API and we can't add fields freely without breaking the binary compatibility.

**Comment by 3flex** [comment]
> it should be public for the integrators

What are the use cases?

I agree with earlier comments that the serialisation should be internal. Exposing results publicly can be done through reports.

**Comment by BraisGabin** [comment]
> What are the use cases?

WOW! The initial description of this issue was REALLY bad. I updated it. Now I think that the use cases are clear. It also helped me to realized that yes, this should be internal. Inside `:detekt-gradle-plugin` because that's the only one that will use it.

Maybe a plugin for bazel or any other tool wants to use it too. But in that case the maintainer of that plugin has 3 options:
- Build it from scratch
- Copy&paste
- Raise an issue so we move that code to `:detekt-tooling`.

In any case we don't need to make it public right now.

**Comment by 3flex** [comment]
What's left for the specific goal of making output serializable? Sub-issues are complete. This would also help with merging baselines (#1589) - I'm going to mark that blocked by this.

**Comment by BraisGabin** [comment]
There is nothing left. It should be doable now. I didn't start with it because right now I have a lot of open PRs. Some waiting for review and others waiting for the others to be merged first. I'm generating a lot of conflicts with myself lately and I have a lot of open fronts. I want to close some before I open a new one.

That said, if anyone wants to go for it and implement this feature that would be great :)
